File: Leetcode875/Leetcode875.py
from math import ceil

class Solution:
    def minEatingSpeed(self, piles, h) -> int:
        
        def eatK_perH(k):
            total_hour = 0
            for pile in piles:
                total_hour += ceil(pile/k)
                if total_hour > h:
                    return False
            return True

        left = 1
        right = max(piles)
        while left <= right:
            mid = (left + right) // 2
            if eatK_perH(mid):
                right = mid - 1
            else:
                left = mid + 1
        return left
    
Solution = Solution()
cases = {"case1":([30,11,23,4,20], 5),
         "case2":([30,11,23,4,20], 6),
         "case3":([3,6,7,11], 8)}
for case in cases.values():
    print(Solution.minEatingSpeed(case[0], case[1]))

File: Leetcode875/Leetcode875.java
package Leetcode875;
import java.util.*;

class Leetcode875 {
    public static void main(String[] args) {
        int[] piles = new int[]{30, 11, 23, 4, 20};
        int h = 5;
        int res = Solution.minEatingSpeed(piles, h);
        System.out.println(res);
    }
}


class Solution {
    public static int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = Arrays.stream(piles).max().orElse(1); // maximum pile

        while (left < right) {
            int mid = (left + right) / 2;

            if (canFinish(piles, h, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    private static boolean canFinish(int[] piles, int h, int k) {
        int hours = 0;
        for (int pile : piles) {
            hours += (pile + k - 1) / k; // Equivalent to ceil(pile / k)
        }
        return hours <= h;
    }
}
